removed 0.8 argument, which ImageMath does not support

richardbeare opened the pull request:
Errors are being caused by ImageMath MajorityVote 0.8 image1 image2 etc.

ImageMath MajorityVote doesn't take a scalar...

ntustison commented:
This is actually a "hidden usage" for ``MajorityVoting``.  See [here](https://github.com/ANTsX/ANTs/blob/master/Examples/ImageMath_Templates.hxx#L11460-L11464).  Is it failing or is it simply declaring that the file "0.8" doesn't exist?

richardbeare commented:
Appears to be doing both - complains then segfault ...
 file 0.8 does not exist .
/opt/slurm/var/spool/job1807776/slurm_script: line 2:  3987 Segmentation
fault      (core dumped) /usr/local/ants/2.2.0/bin//ImageMath 3
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/TargetMaskImageMajorityVoting.nii.gz
MajorityVoting 0.8
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1001_3_BrainCerebellum_1_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1002_3_BrainCerebellum_2_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1006_3_BrainCerebellum_3_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1007_3_BrainCerebellum_4_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1008_3_BrainCerebellum_5_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1009_3_BrainCerebellum_6_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1010_3_BrainCerebellum_7_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1011_3_BrainCerebellum_8_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1012_3_BrainCerebellum_9_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1013_3_BrainCerebellum_10_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1014_3_BrainCerebellum_11_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1015_3_BrainCerebellum_12_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1017_3_BrainCerebellum_13_WarpedLabels.nii.gz
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1036_3_BrainCerebellum_14_WarpedLabels.nii.gz


On Tue, Mar 27, 2018 at 3:11 PM, Nick Tustison <notifications@github.com>
wrote:

> This is actually a "hidden usage" for MajorityVoting. See here
> <https://github.com/ANTsX/ANTs/blob/master/Examples/ImageMath_Templates.hxx#L11460-L11464>.
> Is it failing or is it simply declaring that the file "0.8" doesn't exist?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ANTsX/ANTs/pull/556#issuecomment-376391198>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAvoocgiSx4vqdivojy3KsPke6_kDEygks5tibvqgaJpZM4S8QVj>
> .
>

ntustison commented:
When was the last time you updated your ANTs repo?

richardbeare commented:
The problem version was labelled as 2.2 (installed by cluster admin). I
updated my copy yesterday. Last nights test failed due to other reasons,
but hopefully I'll know today whether it fixes the problem.

On Wed, Mar 28, 2018 at 12:26 AM, Nick Tustison <notifications@github.com>
wrote:

> When was the last time you updated your ANTs repo?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ANTsX/ANTs/pull/556#issuecomment-376524029>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAvooQ33qEeVJY22XUaiveuODEBZZPxKks5tij4EgaJpZM4S8QVj>
> .
>

richardbeare commented:
I've cleaned, updated and rebuilt. I'm running serially to avoid job
submission problems.

I now have errors in antsRegistration, from the following:

/home/rbeare/cv41_scratch/Software/ANTS/bin//antsRegistrationSyNQuick.sh
-d 3                           -p d                           -j
1                           -t s                           -f
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz
-m
/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz
-o
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_
>
/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_log.txt


ERROR:  The number of shrink factors does not match the number of levels.
Shrink Factors: 4
Number Of Levels: 1


On Wed, Mar 28, 2018 at 12:26 AM, Nick Tustison <notifications@github.com>
wrote:

> When was the last time you updated your ANTs repo?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ANTsX/ANTs/pull/556#issuecomment-376524029>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAvooQ33qEeVJY22XUaiveuODEBZZPxKks5tij4EgaJpZM4S8QVj>
> .
>

ntustison commented:
The actual ``antsRegistration`` call gets dumped to the screen right after listing the parameters.  Can you post that call?

richardbeare commented:
From the log:

/home/rbeare/cv41_scratch/Software/ANTS/bin//antsRegistration --verbose 1
--dimensionality 3 --float 0 --collapse-output-transforms 1 --output
[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_,/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_Warped.nii.gz,/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_InverseWarped.nii.gz]
--interpolation Linear --use-histogram-matching 1
--winsorize-image-intensities [0.005,0.995] --initial-moving-transform
[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1]
--transform Rigid[0.1] --metric
MI[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1,32,Regular,0.25]
--convergence 1 --shrink-factors 8x4x2x1 --smoothing-sigmas 3x2x1x0vox
--transform Affine[0.1] --metric
MI[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1,32,Regular,0.25]
--convergence 1 --shrink-factors 8x4x2x1 --smoothing-sigmas 3x2x1x0vox
--transform SyN[0.1,3,0] --metric
MI[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1,32]
--convergence 1 --shrink-factors 8x4x2x1 --smoothing-sigmas 3x2x1x0vox


On Wed, Mar 28, 2018 at 1:06 PM, Nick Tustison <notifications@github.com>
wrote:

> The actual antsRegistration call gets dumped to the screen right after
> listing the parameters. Can you post that call?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ANTsX/ANTs/pull/556#issuecomment-376735733>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAvooXVt2bry6-p4Z72MJLQl_dOGL6Uoks5tivAUgaJpZM4S8QVj>
> .
>

ntustison commented:
Could be a shell thing. See [here](https://github.com/ANTsX/ANTs/blob/master/Scripts/antsRegistrationSyNQuick.sh#L428-L453) that all the convergence parameters are explicitly specified and shouldn't equal '1' but somehow that is what you're getting.   I've seen this before when somebody was using a shell other than bash.

cookpa commented:
tcsh will do this - it will evaluate the parameters in square brackets, replacing them with 1.

The shebang at the top of the script ought to call bash, but it will only work if `/bin/bash` exists and is actually bash (try `/bin/bash --version`).

richardbeare commented:
Ah - the qscript, submitted, created by antsJointLabelFusion, uses /bin/sh

https://github.com/ANTsX/ANTs/blob/b7102451a80db77e30a91fe31745ba523fae3290/Scripts/antsJointLabelFusion.sh#L597



On Wed, Mar 28, 2018 at 1:27 PM, Nick Tustison <notifications@github.com>
wrote:

> Could be a shell thing. See here
> <https://github.com/ANTsX/ANTs/blob/master/Scripts/antsRegistrationSyNQuick.sh#L428-L453>
> that all the convergence parameters are explicitly specified and shouldn't
> equal '1' but somehow that is what you're getting. I've seen this before
> when somebody was using a shell other than bash.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ANTsX/ANTs/pull/556#issuecomment-376739330>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAvooW0xddwVITRa0rofMGWZ44r6gcx3ks5tivUsgaJpZM4S8QVj>
> .
>

cookpa commented:
If you are running `antsJointLabelFusion.sh` serially, then the qscript should be run in bash:

https://github.com/ANTsX/ANTs/blob/7372d027ff29c961578bc286e899618d3bf1a2e7/Scripts/antsJointLabelFusion.sh#L638

In either case, the qscript should in turn execute `antsRegistrationSyNQuick.sh`, so its shebang should take over.

richardbeare commented:
Yes indeed, something very strange going on.... Tracing through a debug
dump now.


On Wed, Mar 28, 2018 at 1:52 PM, Philip Cook <notifications@github.com>
wrote:

> If you are running antsJointLabelFusion.sh serially, then the qscript
> should be run in bash:
>
> https://github.com/ANTsX/ANTs/blob/7372d027ff29c961578bc286e89961
> 8d3bf1a2e7/Scripts/antsJointLabelFusion.sh#L638
>
> In either case, the qscript should in turn execute
> antsRegistrationSyNQuick.sh, so its shebang should take over.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ANTsX/ANTs/pull/556#issuecomment-376743610>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAvooSmNQ-AAsby4V_6SOAI2_DwqtIB0ks5tivsBgaJpZM4S8QVj>
> .
>

richardbeare commented:
Building the command line works correctly - the debug output shows a
sensible convergence value, but those values get replaced by 1:

echo $BASH_VERSION
4.2.46(1)-release


+ PRECISION='--float 0'
+ COMMAND='/home/rbeare/cv41_scratch/Software/ANTS/bin//antsRegistration
--verbose 1                  --dimensionality 3 --float 0
--collapse-output-transforms 1                  --output
[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_,/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_Warped.nii.gz,/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_InverseWarped.nii.gz]
--interpolation Linear                  --use-histogram-matching
1                  --winsorize-image-intensities
[0.005,0.995]                                    --initial-moving-transform
[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1]
--transform Rigid[0.1]             --metric
MI[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1,32,Regular,0.25]
--convergence [1000x500x250x0,1e-6,10]             --shrink-factors
8x4x2x1             --smoothing-sigmas 3x2x1x0vox --transform
Affine[0.1]              --metric
MI[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1,32,Regular,0.25]
--convergence [1000x500x250x0,1e-6,10]              --shrink-factors
8x4x2x1              --smoothing-sigmas 3x2x1x0vox --transform
SyN[0.1,3,0]               --metric
MI[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1,32]
--convergence [100x70x50x0,1e-6,10]           --shrink-factors
8x4x2x1           --smoothing-sigmas 3x2x1x0vox'
+ echo ' antsRegistration call:'
+ echo
--------------------------------------------------------------------------------------
+ echo /home/rbeare/cv41_scratch/Software/ANTS/bin//antsRegistration
--verbose 1 --dimensionality 3 --float 0 --collapse-output-transforms 1
--output
'[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_,/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_Warped.nii.gz,/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/antsCT//malf/1000_3_BrainCerebellum_0_InverseWarped.nii.gz]'
--interpolation Linear --use-histogram-matching 1
--winsorize-image-intensities '[0.005,0.995]' --initial-moving-transform
'[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1]'
--transform 'Rigid[0.1]' --metric
'MI[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1,32,Regular,0.25]'
--convergence 1 --shrink-factors 8x4x2x1 --smoothing-sigmas 3x2x1x0vox
--transform 'Affine[0.1]' --metric
'MI[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1,32,Regular,0.25]'
--convergence 1 --shrink-factors 8x4x2x1 --smoothing-sigmas 3x2x1x0vox
--transform 'SyN[0.1,3,0]' --metric
'MI[/home/rbeare/cv41_scratch/ADNI/ANTs/Raw/ADNIbrain.nii.gz,/home/rbeare/cv41_scratch/ADNI/MICCAI2012-Multi-Atlas-Challenge-Data//training-images/1000_3_BrainCerebellum.nii.gz,1,32]'
--convergence 1 --shrink-factors 8x4x2x1 --smoothing-sigmas 3x2x1x0vox


On Wed, Mar 28, 2018 at 1:52 PM, Philip Cook <notifications@github.com>
wrote:

> If you are running antsJointLabelFusion.sh serially, then the qscript
> should be run in bash:
>
> https://github.com/ANTsX/ANTs/blob/7372d027ff29c961578bc286e89961
> 8d3bf1a2e7/Scripts/antsJointLabelFusion.sh#L638
>
> In either case, the qscript should in turn execute
> antsRegistrationSyNQuick.sh, so its shebang should take over.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ANTsX/ANTs/pull/556#issuecomment-376743610>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAvooSmNQ-AAsby4V_6SOAI2_DwqtIB0ks5tivsBgaJpZM4S8QVj>
> .
>

richardbeare commented:
Seems that on my desktop system (linux mint, bash 4.3.48) the bracket behavior is as follows:

```bash
x=[100x70x50x0,1e-6,10]
echo $x
[100x70x50x0,1e-6,10]
```

While on the cluster, which has older, CentOS, bash 4.2.46, the behavior is:

```bash
x=[100x70x50x0,1e-6,10]
echo $x
1
```
but putting in spaces appears to fix things (I'll check more carefully in a bit):
```bash 
x="[ 100x70x50x0,1e-6,10 ]"
echo $x
[ 100x70x50x0,1e-6,10 ]
```

richardbeare commented:
The process appears to have completed successfully after the spaces were included.

ntustison commented:
Thanks @richardbeare .  Good to know.
